analyzers/src/SonarAnalyzer.Common/Rules/ParametersCorrectOrderBase.cs

```diff
@@ -18,141 +18,161 @@
 
-        protected abstract TypeInfo GetArgumentTypeSymbolInfo(TArgumentSyntax argument, SemanticModel semanticModel);
-        protected abstract Location GetMethodDeclarationIdentifierLocation(SyntaxNode syntaxNode);
-        protected abstract SyntaxToken? GetArgumentIdentifier(TArgumentSyntax argument);
-        protected abstract SyntaxToken? GetNameColonArgumentIdentifier(TArgumentSyntax argument);
+    private readonly DiagnosticDescriptor rule;
 
-        internal void ReportIncorrectlyOrderedParameters(SonarSyntaxNodeReportingContext analysisContext,
-            MethodParameterLookupBase<TArgumentSyntax> methodParameterLookup, SeparatedSyntaxList<TArgumentSyntax> argumentList,
-            Func<Location> getLocationToReport)
-        {
-            var argumentParameterMappings = methodParameterLookup.GetAllArgumentParameterMappings()
-                .ToDictionary(pair => pair.Node, pair => pair.Symbol);
+    protected abstract TypeInfo GetArgumentTypeSymbolInfo(TArgumentSyntax argument, SemanticModel model);
+    protected abstract Location GetMethodDeclarationIdentifierLocation(SyntaxNode syntaxNode);
+    protected abstract SyntaxToken? GetArgumentIdentifier(TArgumentSyntax argument, SemanticModel model);
+    protected abstract SyntaxToken? GetNameColonArgumentIdentifier(TArgumentSyntax argument);
 
-            var methodSymbol = methodParameterLookup.MethodSymbol;
-            if (methodSymbol == null)
-            {
-                return;
-            }
+    protected ParametersCorrectOrderBase() : base(DiagnosticId) =>
+        rule = Language.CreateDescriptor(DiagnosticId, MessageFormat);
 
-            var parameterNames = argumentParameterMappings.Values
-                .Select(symbol => symbol.Name)
-                .Distinct()
-                .ToList();
-
-            var argumentIdentifiers = argumentList
-                .Select(argument => ConvertToArgumentIdentifier(argument))
-                .ToList();
-            var identifierNames = argumentIdentifiers
-                .Select(p => p.IdentifierName)
-                .ToList();
-
-            if (parameterNames.Intersect(identifierNames).Any() &&
-                HasIncorrectlyOrderedParameters(argumentIdentifiers, argumentParameterMappings, parameterNames, identifierNames,
-                    analysisContext.SemanticModel))
-            {
-                // for VB the symbol does not contain the method syntax reference
-                var secondaryLocations = methodSymbol.DeclaringSyntaxReferences
-                    .Select(s => GetMethodDeclarationIdentifierLocation(s.GetSyntax()))
-                    .WhereNotNull();
+    internal void ReportIncorrectlyOrderedParameters(SonarSyntaxNodeReportingContext analysisContext,
+        MethodParameterLookupBase<TArgumentSyntax> methodParameterLookup,
+        SeparatedSyntaxList<TArgumentSyntax> argumentList,
+        Func<Location> getLocationToReport)
```

Comment:
This delegate makes the logic very hard and complex. Send the reporting node instead and call `reportingNode.GetLocation()` below. Then, you won't need any of those extracted local functions and you could simply call it with embedded ternary operator
```suggestion
        Node reportingNode)
```

Reply:
Yep, this is so much better, thanks!